_posts/2020-07-28-state-of-the-art-nlp-models-from-r/state-of-the-art-nlp-models-from-r.Rmd

```diff
@@ -0,0 +1,283 @@
+      
+      txt = tokenizer$encode(data[['comment_text']][i],max_length = max_len, 
+                             truncation=T) %>% 
+        t() %>% 
+        as.matrix() %>% list()
+      lbl = data[['target']][i] %>% t()
+      
+      text = text %>% append(txt)
+      label = label %>% append(lbl)
+    }
+    list(do.call(plyr::rbind.fill.matrix,text), do.call(plyr::rbind.fill.matrix,label))
+  }
+  
+  train_ = data_prep(train)
+  test_ = data_prep(test)
+  
+  # slice dataset
+  tf_train = tensor_slices_dataset(list(train_[[1]],train_[[2]])) %>% 
+    dataset_batch(batch_size = batch_size, drop_remainder = TRUE) %>% 
+    dataset_shuffle(128) %>% dataset_repeat(epochs) %>% 
+    dataset_prefetch(tf$data$experimental$AUTOTUNE)
+  
+  tf_test = tensor_slices_dataset(list(test_[[1]],test_[[2]])) %>% 
+    dataset_batch(batch_size = batch_size)
+  
+  # create an input layer
+  input = layer_input(shape=c(max_len), dtype='int32')
+  hidden_mean = tf$reduce_mean(model_(input)[[1]], axis=1L) %>% 
+    layer_dense(64,activation = 'relu')
+  # create an output layer for binary classification
+  output = hidden_mean %>% layer_dense(units=1, activation='sigmoid')
+  model = keras_model(inputs=input, outputs = output)
+  
+  # compile with AUC score
+  model %>% compile(optimizer= tf$keras$optimizers$Adam(learning_rate=3e-5, epsilon=1e-08, clipnorm=1.0),
+                    loss = tf$losses$BinaryCrossentropy(from_logits=F),
+                    metrics = tf$metrics$AUC())
+  
+  print(glue::glue('{ai_m[[i]][1]}'))
+  # train the model
+  history = model %>% keras::fit(tf_train, epochs=epochs, #steps_per_epoch=len/batch_size,
+                validation_data=tf_test)
+  gather_history[[i]]<- history
+  names(gather_history)[i] = ai_m[[i]][1]
+}
+
+```
+
+<center>
+<a href="https://www.kaggle.com/henry090/transformers" class="colab-root">Reproduce in a<span class="colab-span">  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Notebook</span></a>
+</center>
+<br>
+
+Extract results to see the benchmarks:
+
+```{r eval = F, echo = T}
+res = sapply(1:3, function(x) {
+  do.call(rbind,gather_history[[x]][["metrics"]]) %>% 
+    as.data.frame() %>% 
+    tibble::rownames_to_column() %>% 
+    mutate(model_names = names(gather_history[x])) 
+}, simplify = F) %>% do.call(plyr::rbind.fill,.) %>% 
+  mutate(rowname = stringr::str_extract(rowname, 'loss|val_loss|auc|val_auc')) %>% 
+  rename(epoch_1 = V1, epoch_2 = V2)
+```
+
+```{r eval=T,echo=F}
+library(dplyr)
+res = data.table::fread('files/res.csv') %>% 
+  filter(rowname %in% 'val_auc') %>% arrange(desc(V2)) %>% 
+   rename(epoch_1 = V1, epoch_2 = V2, metric = rowname) %>% 
+  mutate(epoch_1 = round(epoch_1,2),epoch_2 = round(epoch_2,2))
+DT::datatable(res, options = list(dom = 't'))
+```
+
+RoBERTa and Electra models have some improvements after 2 epochs what cannot be said for GPT-2 model. In this case, it is clear that it can be enough to train state-of-the-art model even for 1 epoch.
```

Comment:
```suggestion
Both the _RoBERTa_ and _Electra_ models show some additional improvements after 2 epochs of training, which cannot be said of  _GPT-2_. In this case, it is clear that it can be enough to train a state-of-the-art model even for a single epoch.
```